### PR TITLE
fix(cognito): Identity ID api returns invalid error on concurrent calls

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:* https://github.com/aws-amplify/amplify-ios/issues/1822

*Description of changes:* 

*Check points:*

- [x] Added new tests to cover change, if needed
- [x] All unit tests pass
- [ ] All integration tests pass
- [ ] Updated CHANGELOG.md
- [x] Documentation update for the change if required
- [x] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AWSCore/Authentication/AWSIdentityProvider.m [hunk 1]
```diff
@@ -292,52 +290,67 @@ - (NSString *)identityProviderName {
...
+    }];
+}
+
+- (AWSTask<NSString *> *)internalGetIdentityId {
+
     if (self.identityId) {
+        AWSDDLogVerbose(@"internalGetIdentityId: Cached present");
         return [AWSTask taskWithResult:self.identityId];
     } else {
+        AWSDDLogVerbose(@"internalGetIdentityId: Fetching");
```
Comment (palpatim) [inline AWSCore/Authentication/AWSIdentityProvider.m:318]: nit 
```suggestion
        AWSDDLogVerbose(@"internalGetIdentityId: Fetching identity ID");
```

File: AWSCore/Authentication/AWSIdentityProvider.m [hunk 2]
```diff
@@ -292,52 +290,67 @@ - (NSString *)identityProviderName {
...
+    return [task continueWithBlock:^id _Nullable(AWSTask * _Nonnull t) {
+        AWSDDLogVerbose(@"getIdentityId: Going to wait %@", task);
+        dispatch_semaphore_wait(self.semaphore, DISPATCH_TIME_FOREVER);
+        AWSDDLogVerbose(@"getIdentityId: Get internal identity %@", task);
+        return [[self internalGetIdentityId]
+                continueWithBlock:^id _Nullable(AWSTask<NSString *> * _Nonnull t) {
+            AWSDDLogVerbose(@"getIdentityId: Going to signal %@", task);
+            dispatch_semaphore_signal(self.semaphore);
+            return t;
+        }];
```
[the remark this answers is not in the document]
Comment (palpatim) [reply AWSCore/Authentication/AWSIdentityProvider.m:308]: Continuing thoughts about edge cases:
1. Now that we no longer have a 5 second timeout, what happens if we kick off this process and the app goes to background? 
2. What if the task is cancelled mid-flight?

**Comment by royjit** [comment]
Need to rework on this PR, I am getting into deadlock if I try a large number of getIdentityId just after signIn
```swift
for x in 1...100 {
                    DispatchQueue.global().async {
                        AWSMobileClient.default().getIdentityId().continueWith(block: { task in
                            
                        })
                    }
                }
```

**Comment by royjit** [comment]
> Need to rework on this PR, I am getting into deadlock if I try a large number of getIdentityId just after signIn
> 
> ```swift
> for x in 1...100 {
>                     DispatchQueue.global().async {
>                         AWSMobileClient.default().getIdentityId().continueWith(block: { task in
>                             
>                         })
>                     }
>                 }
> ```

Closing this PR, and raised a different one to handle the issue. https://github.com/aws-amplify/aws-sdk-ios/pull/4202

**Pull request closed without merging**